### ISSUE TITLE
Is Reweighing doing what is expected to do?

### ISSUE THREAD

**Issue by salimnoma** [Trusted-AI/AIF360#429]
Hi,

I noticed the reweighing mitigation algorithm is only learning 4 different sample weights (p/f, up/f, p/uf, up/uf):

```
# reweighing weights
   self.w_p_fav = n_fav*n_p / (n*n_p_fav)
   self.w_p_unfav = n_unfav*n_p / (n*n_p_unfav)
   self.w_up_fav = n_fav*n_up / (n*n_up_fav)
   self.w_up_unfav = n_unfav*n_up / (n*n_up_unfav)
```

is this the expected behavior?

From the AIF360's documentation itself and when reviewing the cited papers, it seems like the appropriate behavior is to learn a separate weight for each group and class combination. 

Docs:

> Reweighing is a preprocessing technique that Weights the examples in each (group, label) combination differently to ensure fairness before classification 

Cited Paper:
![Screen Shot 2022-12-29 at 2 54 25 PM](https://user-images.githubusercontent.com/75376289/210005329-1c3f4d60-a986-4021-a770-df14f585defc.png)

![Screen Shot 2022-12-29 at 2 54 37 PM](https://user-images.githubusercontent.com/75376289/210005340-2b9176f8-74aa-4bf1-9342-9dd2abeb7278.png)

**Comment by kvarsh** [comment]
Yes this is the correct behavior. A binary classification problem with one privileged group and one unprivileged group has 4 total combinations.

**Comment by salimnoma** [comment]
But the class parameters are expected to support a list. So either groups can have more than one member:

```
 unprivileged_groups (list(dict)): Representation for unprivileged group.
 privileged_groups (list(dict)): Representation for privileged group.
``` 

On line 48, the code is actually using the input privileged groups and merges them to obtain p/up condtions:

```
cond_p_fav, cond_p_unfav, cond_up_fav, cond_up_unfav) =\
                self._obtain_conditionings(dataset)
```

**Comment by salimnoma** [comment]
Also, it seems generally helpful to be able to allow heterogeneous sample weight assignment for sub populations based on the protected attributes. For example race. I noticed after using reweighing the largest unprivileged group was improved significantly while the mitigation had little or not impact. Wouldn't be helpful to be determine multiple sub-groups of the unprivileged group and allow the sample weights to be tunes to each group? This would be more closely aligned to the cited algorithm as well.

I ended up implementing it for myself. It would be easy to add the functionality to the class allowing people to choose between the mode.

**Comment by hoffmansc** [comment]
Have you checked out the 'aif360.sklearn.preprocessing' version of Reweighing? That version should work like you expect.

This version only works with binary groups: one privileged and one unprivileged. The reason they accept lists is so you can arbitrarily assign subgroups to be either privileged or unprivileged.

**Comment by salimnoma** [comment]
I implemented it internally but 'aif360.sklearn.preprocessing' sounds useful for the future. I didn't know everything under the default algorithms assumes that we have one privileged and unprivileged group. Thanks for the clarification. I'll close the issue.
